### PR TITLE
bh/lazy loading

### PR DESCRIPTION
**Pull request by bheni**

### CODE REVIEW

File: sql/core.go [hunk 1]
```diff
@@ -283,8 +285,57 @@ type RowUpdater interface {
 // Database represents the database.
 type Database interface {
 	Nameable
-	// Tables returns the information of all tables.
-	Tables() map[string]Table
+
+	// GetTableInsensitive retrieves a table by it's name where capitalization does not matter
+	GetTableInsensitive(ctx context.Context, tblName string) (Table, bool, error)
```
Comment (zachmu) [inline sql/core.go]: Spell out how the case sensitivity should work: prefer an exact match, fall back to single case insensitive match if there isn't a case sensitive match, error if there is more than one case-insensitive match. Probably mention the GetTableInsensitive method below. Also mention the two non-obvious return params.

File: sql/core.go [hunk 2]
```diff
@@ -283,8 +285,57 @@ type RowUpdater interface {
...
+
+	// GetTableNames returns the table names of every table in the database
+	GetTableNames(ctx context.Context) ([]string, error)
+}
+
+// GetTableIncensitive implements a case insensitive map lookup for tables keyed off of the table name
+func GetTableInsensitive(tblName string, tables map[string]Table) (Table, bool) {
```
Comment (zachmu) [inline sql/core.go]: This should look for an exact match before looking for a case insensitive match.

File: sql/core.go [hunk 3]
```diff
@@ -283,8 +285,57 @@ type RowUpdater interface {
...
+	}
+
+	tbl, ok := lwrMap[strings.ToLower(tblName)]
+
+	return tbl, ok
+}
+
+// DBTableIter iterates over all tables returned by db.GetTableNames() calling cb for each one until all tables have
+// been processed, or an error is returned from the callback, or the stop flag is returned from the callback.
+func DBTableIter(ctx context.Context, db Database, cb func(Table) (stop bool, err error)) error {
```
Comment (zachmu) [inline sql/core.go]: The rest of this code bases uses continue flags, not stop flags. You should follow that convention

File: sql/index.go [hunk 4]
```diff
@@ -215,61 +222,109 @@ func (r *IndexRegistry) RegisterIndexDriver(driver IndexDriver) {
...
+
+			if err != nil {
+				return err
+			}
+
+			for tIdx := range tNames {
+				tName := tNames[tIdx]
+
+				loadF := func(ctx context.Context) error {
+					t, ok, err := db.GetTableInsensitive(ctx, tName)
```
Comment (zachmu) [inline sql/index.go]: You need to deal with the lock semantics in this function scope instead of in the parent scope

File: sql/core.go (hunk 2 above)
Comment (zachmu) [reply sql/core.go]: It should also throw an error if there is more than one case-insensitive match.

File: sql/core.go (hunk 1 above)
Comment (bheni) [reply sql/core.go]: updated.

File: sql/core.go (hunk 2 above)
Comment (bheni) [reply sql/core.go]: updated.  added a comment that if there is more than one case insensitive match that the one that is returned is not specified.

File: sql/core.go (hunk 3 above)
Comment (bheni) [reply sql/core.go]: updated

Comment (bheni) [reply sql/core.go]: updated

File: sql/index.go (hunk 4 above)
Comment (bheni) [reply sql/index.go]: discussed

**Pull request merged**